src/ILLink.RoslynAnalyzer/DataFlow/LocalStateLattice.cs

```diff
@@ -29,20 +30,34 @@ public override string ToString ()
 		}
 	}
 
-	// Wrapper struct exists purely to substitute a concrete LocalKey for TKey of DefaultValueDictionary
 	public struct LocalState<TValue> : IEquatable<LocalState<TValue>>
 		where TValue : IEquatable<TValue>
 	{
 		public DefaultValueDictionary<LocalKey, TValue> Dictionary;
 
-		public LocalState (DefaultValueDictionary<LocalKey, TValue> dictionary) => Dictionary = dictionary;
+		public DefaultValueDictionary<CaptureId, PropertyValue> CapturedProperties;
```

Comment:
It could work as a SingleValue, if we moved the logic to create/handle PropertyValue over to TrimAnalysisVisitor instead of LocalDataFlowVisitor (which is generic over TValue). But the handling of property assignments is something that would likely be shared by any dataflow analysis, so I'd prefer to handle it in the generic code. Maybe a different way to accomplish that would be to have the visitor return some wrapper for TValue which includes the captured properties?

It's also not really a value in the same way that the other dataflow values are, since it doesn't represent an abstraction of any actual runtime value, at least the way I have been thinking about it.

Reply:
I see - that makes it harder I agree.

I don't see too much of a difference - for me it is a value. The fact that linker has to treat it as method calls makes things more complicated in some sense. But I do agree that tracking the capture IDs should be on the "general" data flow, not on the "trimming specific" code.

I'm fine with it this way, it's a relatively minor difference. And it guarantees that it won't leak through to the shared code, which is nice.